src/App.js

```diff
@@ -97,105 +115,49 @@ let App = () => {
-
-      // What events are currently visible?
-      // Basically the module's full list of classes, minus alternatives to chosen options (from the query string)
-      const eventsForModule = [...timetableData[`${id}_${session}`].classes]
-      for (const [module, groupId, occurrence] of specifiedOccurrences) {
-        if (module !== id) continue
-        // Delete alternatives to an explicitly chosen event
-        for (let i = eventsForModule.length - 1; i >= 0; i--) {
-          const event = eventsForModule[i]
-          if (event.activity === groupId && parseInt(event.occurrence) !== occurrence) {
-            eventsForModule.splice(i, 1)
-          }
-        }
-      }
-      // Delete hidden occurrences
-      for (const [module, groupId, occurrence] of hiddenOccurrences) {
-        if (module !== id) continue
-        for (let i = eventsForModule.length - 1; i >= 0; i--) {
-          const event = eventsForModule[i]
-          if (event.activity === groupId && parseInt(event.occurrence) === occurrence) {
-            eventsForModule.splice(i, 1)
-          }
-        }
-      }
-
-      // Add currently visible events to the calendar
-      api.addEventSource({
-        id,
-        color: stringToColor(id),
-        events: parseEvents(eventsForModule, year, session, id)
-      })
-    })
-  }, [timetableData, year, session, selectedModules, calendar, timeZone, m, modules, specifiedOccurrences, hiddenOccurrences])
 
-  // Remove specified events for modules that have been removed
-  useEffect(() => {
-    updateSpecifiedOccurrences()
-    updateHiddenOccurrences()
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [selectedModules])
-
-  // We select occurrences by adding them to specifiedOccurrences, which
-  // edits the query string in an effect
-  const selectOccurrence = (module, groupId, occurrence) => {
-    // Eg adding ['COMP1130', 'ComA', 1]
-    setSpecifiedOccurrences([...specifiedOccurrences, [module, groupId, occurrence]])
-  }
-  const resetOccurrence = (module, groupId, occurrence) => {
-    setSpecifiedOccurrences(specifiedOccurrences.filter(
-      ([m, g, o]) => !(m === module && g === groupId && o === occurrence)
-    ))
+  const changeOccurrences = (state, action) => {
+    let [module, groupId, occurrence] = action.values
+    switch (action.type) {
+      case 'select':
+        setQueryParam(module, groupId + occurrence)
+        return [...state, action.values]
+      case 'reset':
+        setQueryParam(module, '')
+        return state.filter(
+          ([m, g, o]) => !(m === module && g === groupId && o === occurrence)
+        )
+      default:
+        throw new Error()
+    }
   }
-  const hideOccurrence = (module, groupId, occurrence) => {
-    setHiddenOccurrences([...hiddenOccurrences, [module, groupId, occurrence]])
+
+  const [specifiedOccurrences, setSpecifiedOccurrences] = useReducer(changeOccurrences, getSpecOccurrences())
+
+  const changeHidden = (state, action) => {
```

Comment:
I'm not experienced with state management, why specify the action here vs determining it via size of `action.values`?

Reply:
I think it just makes the `dispatch` call site clearer, but you're right that we could get away without it

Reply:
ah. wishing I'd used a proper hook for all this, would've been as simple as

```javascript
const [hidden, setHidden] = useQueryParam(h, DelimitedArrayParam)
setHidden(hidden => ([...hidden, values])) // hide
setHidden([]) // reset
```

Reply:
Converted back to useState as the reducer made it more complex than necessary.